sandbox.go

```diff
@@ -1,130 +1,317 @@
+	"github.com/docker/docker/client"
+	uuid "github.com/satori/go.uuid"
+)
+
+const (
+	TmpDirPrefix = "xaqt-"
 )
 
+// prepares for execution of user code by creating a temp directory for
+// code and input.
+//
 type sandbox struct {
+	// sandbox id (uuidV4)
+	ID       string
 	language ExecutionDetails
 	code     string
 	stdin    string
 	options  options
+	// docker client connection
+	docker *client.Client
+	// wait channel for successful container exit
+	waitChan <-chan container.ContainerWaitOKBody
+	// error channel for container error
+	errChan <-chan error
 }
 
-func newSandbox(l ExecutionDetails, code, stdin string, opts options) *sandbox {
-	return &sandbox{l, code, stdin, opts}
+// constructs a new sandbox given...
+//
+func newSandbox(l ExecutionDetails, code, stdin string, opts options) (*sandbox, error) {
+	var (
+		s   *sandbox
+		err error
+	)
+
+	// set the API version to use in an environment variable
+	// TODO it would be nice to configure based on the docker version
+	// a user currently has.... not enough time right now so skipping that.
+	err = os.Setenv("DOCKER_API_VERSION", "1.35")
+	if err != nil {
+		return nil, err
+	}
+
+	// init a docker api client
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		// this could occur if docker has not been installed or started
+		return nil, err
+	}
+
+	// TODO (cw|4.29.2018) if we are spinning up this sandbox from within another docker
+	// container, we may want to define a bridge network between them (since they will be
+	// sibling containers). I don't know if this is entirely necessary though...
+	// THIS NETWORK SETUP SHOULD ACTUALLY GO IN A HIGHER SCOPE (within the struct which
+	// actually constructs sandboxes). this way we aren't creating and destroying docker
+	// networks all over the place. instead we should check to see if one has been created.
+
+	// define unique network name
+	// networkName := fmt.Sprintf("xaqt.%s", uuid.NewV4().String())
+
+	// setup container bridge network if one doesn't already exist.
+	// _, err = dockerClient.NetworkCreate(
+	// 	context.TODO(),
+	// 	networkName,
+	// 	types.NetworkCreate{},
+	// )
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	s = &sandbox{
+		ID:       uuid.NewV4().String(),
+		language: l,
+		code:     code,
+		stdin:    stdin,
+		options:  opts,
+		docker:   dockerClient,
+	}
+
+	return s, nil
 }
 
+// runs user code within the sandbox after preparing the execution environment.
+//
 func (s *sandbox) run() (string, error) {
-	s.prepare()
-	return s.execute()
+	var (
+		output string
+		err    error
+	)
+
+	err = s.prepare()
+	if err != nil {
+		return "", err
+	}
+
+	output, err = s.execute()
+	if err != nil {
+		return "", err
+	}
+
+	return output, nil
 }
 
-func (s *sandbox) prepare() {
-	tmpFolder, err := ioutil.TempDir(s.options.folder, "xaqt-")
+// prepares the execution environment and the sandbox docker container.
+//
+func (s *sandbox) prepare() error {
```

Comment:
Some important cleanup such that we deal with errors more robustly.

Reply:
:+1: